apps/OboeTester/app/src/main/java/com/mobileer/oboetester/CommunicationDeviceView.java

```diff
@@ -0,0 +1,135 @@
+                .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+        inflater.inflate(R.layout.comm_device_view, this);
+
+        mAudioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
+        mSpeakerphoneCheckbox = (CheckBox) findViewById(R.id.setSpeakerphoneOn);
+
+        mDeviceSpinner = (CommunicationDeviceSpinner) findViewById(R.id.comm_devices_spinner);
+        mDeviceSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
+            public void onItemSelected(AdapterView<?> parent, View view, int position, long id)
+            {
+                AudioDeviceInfo[] commDeviceArray = mDeviceSpinner.getCommunicationsDevices();
+                if (commDeviceArray != null) {
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                        if (position == 0) {
+                            mAudioManager.clearCommunicationDevice();
+                        } else {
+                            AudioDeviceInfo selectedDevice = commDeviceArray[position - 1]; // skip "Clear"
+                            mAudioManager.setCommunicationDevice(selectedDevice);
+                        }
+                        showCommDeviceStatus();
+                    }
+                }
+            }
+            public void onNothingSelected(AdapterView<?> parent) {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                    mAudioManager.clearCommunicationDevice();
+                }
+                showCommDeviceStatus();
+            }
+        });
+
+        mSpeakerphoneCheckbox.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                onSetSpeakerphoneOn(view);
+            }
+        });
+        mIsSpeakerphoneText = (TextView) findViewById(R.id.isSpeakerphoneOn);
+        showCommDeviceStatus();
+    }
+
+    public void cleanup() {
+        mSpeakerphoneCheckbox.setChecked(false);
+        setSpeakerPhoneOn(false);
+    }
+
+    public void onSetSpeakerphoneOn(View view) {
+        Log.d(TestAudioActivity.TAG, "onSetSpeakerphoneOn() called from Checkbox");
+        CheckBox checkBox = (CheckBox) view;
+        boolean enabled = checkBox.isChecked();
+        setSpeakerPhoneOn(enabled);
+        showCommDeviceStatus();
+    }
+
+    private void setSpeakerPhoneOn(boolean enabled) {
+        Log.d(TestAudioActivity.TAG, "call setSpeakerphoneOn(" + enabled + ")");
+        mAudioManager.setSpeakerphoneOn(enabled);
+    }
+
+    private void showCommDeviceStatus() {
+        boolean enabled = mAudioManager.isSpeakerphoneOn();
+        String text = (enabled ? "ON" : "OFF");
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
+            AudioDeviceInfo commDeviceInfo = mAudioManager.getCommunicationDevice();
+            if (commDeviceInfo != null) {
+                text += ", CD#" + commDeviceInfo.getId();
```

Comment:
nit: CD seems quite vague. Can we use a longer name?

Reply:
I changed it to "CommDev=". More clear but still fits on the small screen.